src/libexpr/primops.cc

```diff
@@ -1387,6 +1436,31 @@ static void prim_mapAttrs(EvalState & state, const Pos & pos, Value * * args, Va
 }
 
 
+/* Given a set of attribute names, return the set of the corresponding
+   attributes from the given set. Former `lib.genAttrs'
```

Comment:
Let's not get ahead of ourselves ;)

```suggestion
   attributes from the given set. Implements Nixpkgs `lib.genAttrs'
```

-----


`genAttrs` might be a good candidate for a primop regardless of your original goal. As currently implemented it allocates an unnecessary list, and 2-attr attrsets for each attribute name. I don't know what would be the impact of this optimization though. (TODO: register it with a documentation string like most of the other primops, add tests)

Reply:
I would also argue that `builtins` is explicitly not a replacement for `lib`, because the builtins can't be changed. The analogy between kernel syscalls and a standard library (like libc or go) is fitting.